openmicroscopy/bioformats
# Model enum units metadata

This PR is the first part of the C++ units support needed for the 2015 model.  The primary change is the introduction of extended enum metadata to the schema, for use by the generator (it's all `xsd:appinfo/xsdfu` data so isn't a model change).  This allows the enum name to be different from the language enum name, which removes the need to hardcode the differences in the generator.  It also has alternative names to allow for plurals, language differences and alternative regexes, which also removes hardcoded behaviour from the generator and supplementary configuration files (`enum_handler.cfg`).

- Unit language enum names are now the full unit name, matching OMERO.  The symbol names were inconsistent, caused clashes with system headers for C++, and had capitalisation and encoding issues with the micron symbol.  The corresponding code changes are the bulk of this PR.
- Binning names are not hardcoded in the generator; I've chosen a proper name rather than a mixture of names and symbols, which is a breaking change but is a better choice; can be reverted if compatibility is more important
- The code generator passes through a full copy of the parsed schemas, since bugs in generateds strip out needed information like per-enum metadata; this makes much more information available than previously.  Currently it's only used for the enum metadata.
- The C++ generator changes aren't tested here since we don't use the new model yet.  Future work extends (and corrects) some of the changes made here.

--------

Testing:

- Check jobs are green
- Review generated code changes for Java and C++ https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-generated-sources/lastSuccessfulBuild/artifact/ome-xml-model-java.diff
- Verify that the enum name changes are correct and acceptable
- Verify that downstream users are not affected by the changes; if they are then they will need updating

--------

Followup work

- If the Binning and FontFamily changes to ome.xsd are acceptable, I'd like to move all enum definitions to the top-level like this for consistency with the code generator--it would mean only a single method of handling enums in the generator which would remove some unnecessary complexity from it
- Additional C++ PRs for MapAnnotations and 2015 model generation
